lightstep/legacy_query_equivalence.go

```diff
@@ -0,0 +1,151 @@
+package lightstep
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/lightstep/terraform-provider-lightstep/client"
+)
+
+// Checks if the prior dashboard attributes have legacy queries that
+// map to the UQL queries that the API call returns.
+//
+// If so, we can ignore the "false diff" a Terraform plan would otherwise show
+// by using the legacy resource data instread of the UQL resource data.
+func dashboardHasEquivalentLegacyQueries(
+	ctx context.Context,
+	c *client.Client,
+	projectName string,
+	priorAttrs *client.UnifiedDashboardAttributes,
+	updatedAttrs *client.UnifiedDashboardAttributes,
+) (bool, error) {
+	// This code is only applicalbe if there are legacy charts
+	for _, chart := range priorAttrs.Charts {
+		for _, query := range chart.MetricQueries {
+			if query.Type == "tql" {
+				return false, nil
+			}
+		}
+	}
+
+	// Loop through each chart and compare the queries...
+	for index, priorChart := range priorAttrs.Charts {
+		// Assumes the order of charts is _not_ going to change in the internal
+		// data structure.  Note that we can't do an Chart.ID look up since the
+		// prior structure doesn't necessarily have an ID at this point.
+		updateChart := updatedAttrs.Charts[index]
+		equivalent, err := compareUpdatedLegacyQueries(
+			ctx, c, projectName,
+			priorChart.MetricQueries,
+			updateChart.MetricQueries,
+		)
+		if err != nil {
+			return false, err
+		}
+		if !equivalent {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+// See dashboardHasEquivalentLegacyQueries: this is the metric condition
+// version
+func metricConditionHasEquivalentLegacyQueries(
+	ctx context.Context,
+	c *client.Client,
+	projectName string,
+	priorAttrs *client.UnifiedConditionAttributes,
+	updatedAttrs *client.UnifiedConditionAttributes,
+) (bool, error) {
+	// This code is only applicalbe if there are legacy charts
+	for _, query := range priorAttrs.Queries {
+		if query.Type == "tql" {
+			return false, nil
+		}
+	}
+
+	equivalent, err := compareUpdatedLegacyQueries(
+		ctx, c, projectName,
+		priorAttrs.Queries,
+		updatedAttrs.Queries,
+	)
+	if err != nil {
+		return false, err
+	}
+	return equivalent, nil
+}
+
+// Check that the prior and updated set of queries are equivalent by
+// checking if the the UQL translations are the same.
+func compareUpdatedLegacyQueries(
+	ctx context.Context,
+	c *client.Client,
+	projectName string,
+	priorQueries []client.MetricQueryWithAttributes,
+	updatedQueries []client.MetricQueryWithAttributes,
+) (bool, error) {
+
+	// Step 1: call the SaaS to translate the legacy queries to UQL
```

Comment:
TODO: this also needs a "step 0" of comparing all attributes of `MetricQueryWithAttributes` that _do not_ affect the query.  Or alternately, the calling code needs to be updated to only copy the resource data attributes that influence the query itself.

Reply:
slight bias towards the first one; seems clearer than this function requiring some special pre-work in order to be safely callable